### PR TITLE
PlasmaRequestBuffer data structure updates

### PR DESCRIPTION
**Pull request by pschafhalter**
- converted `transfer_queue` to `std::list`
- converted `pending_object_transfers` to `std::unordered_map`

### CODE REVIEW

File: src/plasma/plasma_manager.cc [hunk 1]
```diff
@@ -1446,12 +1439,9 @@ ClientConnection *ClientConnection_init(PlasmaManagerState *state,
                                         int client_sock,
                                         const char *client_key) {
   /* Create a new data connection context per client. */
-  ClientConnection *conn =
-      (ClientConnection *) malloc(sizeof(ClientConnection));
+  ClientConnection *conn = new ClientConnection();
```
Comment (robertnishihara) [inline src/plasma/plasma_manager.cc]: @pschafhalter I believe this was a bug. We forgot to replace the correspodning `free` call with a `delete` (in `ClientConnection_free`).

File: src/plasma/plasma_manager.cc [hunk 2]
```diff
@@ -897,8 +892,7 @@ void process_data_request(event_loop *loop,
                           int64_t data_size,
                           int64_t metadata_size,
                           ClientConnection *conn) {
-  PlasmaRequestBuffer *buf =
-      (PlasmaRequestBuffer *) malloc(sizeof(PlasmaRequestBuffer));
+  PlasmaRequestBuffer *buf = new PlasmaRequestBuffer();
```
Comment (robertnishihara) [inline src/plasma/plasma_manager.cc]: Looks like we introduced a bug here. There were multiple instances where we needed to replace `free` -> `delete`, and we didn't replace all of them.

Comment (pschafhalter) [reply src/plasma/plasma_manager.cc]: Good catch. I doubled checked the code and it looks like the bugs were fixed.

Comment (robertnishihara) [reply src/plasma/plasma_manager.cc]: Yeah, I think I got most of them. It's possible that I missed some. But now that our valgrind tests are passing again, any problems should show up there.

**Pull request merged**